distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Add_Row_Number.enso

```diff
@@ -0,0 +1,103 @@
+from Standard.Base import all
+
+import project.Data.Column.Column
+import project.Data.Column_Selector.Column_Selector
+import project.Data.Sort_Column.Sort_Column
+import project.Data.Set_Mode.Set_Mode
+import project.Data.Table.Table
+import project.Internal.Multi_Value_Key.Unordered_Multi_Value_Key
+import project.Internal.Multi_Value_Key.Ordered_Multi_Value_Key
+import project.Internal.Problem_Builder.Problem_Builder
+import project.Internal.Table_Helpers
+import project.Internal.Unique_Name_Strategy.Unique_Name_Strategy
+
+from project.Errors import Duplicate_Output_Column_Names
+from project.Internal.Java_Exports import make_long_builder
+
+polyglot java import org.enso.base.arrays.LongArray
+polyglot java import org.enso.table.data.column.storage.LongStorage
+
+## PRIVATE
+add_row_number : Table -> Text -> Integer -> Integer -> (Column_Selector | Vector Text) -> Vector (Text | Sort_Column) | Text | Sort_Column -> Problem_Behavior -> Table
+add_row_number table name from step group_by order_by on_problems =
+    problem_builder = Problem_Builder.new error_on_missing_columns=True
+    grouping_columns = table.columns_helper.select_columns_helper group_by True problem_builder
+    Unordered_Multi_Value_Key.validate_grouping_columns grouping_columns problem_builder
+    ordering = Table_Helpers.resolve_order_by table.columns order_by problem_builder
+    problem_builder.attach_problems_before on_problems <|
+        new_column = case ordering.is_empty of
+            True ->
+                case grouping_columns.is_empty of
+                    True -> make_range_column name from step table.row_count
+                    False -> make_grouped_enumeration name grouping_columns from step
+            False -> make_grouped_ordered_enumeration name grouping_columns ordering from step
+
+        column_names = table.column_names
+        renamed_table = if column_names.contains name . not then table else
+            problems = [Duplicate_Output_Column_Names.Error [name]]
```

Comment:
This case needs to be handled a lot; I think this would be a really useful helper method (if we don't have one already).

Reply:
Could you clarify?

Reply:
We are handling duplicate column names, usually directly through `Unique_Name_Strategy`, but this method has slightly different behaviour since it renames the old column while the new column retains its name. Usually it would be the other way round. That's why there is a bit more custom code here.

Reply:
I see, that's a less common case (but I suspect it will come up again at some point). Find to leave it as it is.